### PR TITLE
Allow remediation points to be multipliable

### PR DESCRIPTION
**Pull request by BlakeWilliams**
This adds a check for the format of `[score/limit]` in a cop message and
will extract the `score` value and multiply it by the defined remediation
points.

Remediation points for metric cops were also tweaked to make up for the addition of multiplication.

### CODE REVIEW

File: lib/cc/engine/violation_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+      def cop_list
+        return @cop_list if @cop_list
+
+        cops_path = File.expand_path(
+          File.join(File.dirname(__FILE__), "../../../config/cops.yml")
+        )
+
+        File.open(cops_path, "r") do |file|
+          @cop_list = YAML.load(file.read)
```
Comment (pbrisbin) [inline lib/cc/engine/violation_decorator.rb]: Can we use `SafeYAML` here? Even though it's not user input, better to just always use that everywhere I think. 

Also, I think there's a `.load_file` method (on both modules).

File: lib/cc/engine/violation_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        if multiplier?
+          points * multiply_by
+        else
+          points
+        end
+      end
+
+      private
+
+      def multiply_by
```
Comment (GordonDiggs) [inline lib/cc/engine/violation_decorator.rb]: WDYT of calling this method `multiplier`? Mathematically, that's its function

Comment (GordonDiggs) [inline lib/cc/engine/violation_decorator.rb]: I think the `multipler?` and `multiplier` pattern works well too

File: lib/cc/engine/violation_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        result = message.scan(MULTIPLIER_REGEX)
+        result[0][0].to_i
```
Comment (GordonDiggs) [inline lib/cc/engine/violation_decorator.rb]: Is there a way to get these from the object itself? I assume they have to exist as attributes somewhere for the message to be generated.

File: lib/cc/engine/violation_decorator.rb (hunk 2 above)
Comment (BlakeWilliams) [inline lib/cc/engine/violation_decorator.rb]: Agreed.

File: lib/cc/engine/violation_decorator.rb (hunk 1 above)
Comment (BlakeWilliams) [inline lib/cc/engine/violation_decorator.rb]: I'll move it over.

**Comment by pbrisbin** [comment]
> This adds a check for the format of [score/limit] in a cop message and
will extract the score value and multiply it by the defined remediation
points.

I'm not sure this logic makes sense.

- I'd image scaling by *overage*, not raw *score* is better (`110/100` shouldn't hurt you that much more than `11/10` right?)
- Even then, simple multiplication is probable not right (`11/10` won't impact points in such a case)

I think there is a scaling scheme in legacy analyzer code somewhere that would be closer to what we want:

https://github.com/codeclimate/analyzer/blob/9f96fa2f8c65cc262bedc4a95c5d36091dc7db28/lib/quality/threshold_penalty.rb

This describes things in terms of base-cost, cost-per, and overage. I think base-cost and cost-per are cop-specific and overage could be parsed out of the message.

**Comment by BlakeWilliams** [comment]
That makes sense, I originally had it to only calculate overage but that didn't feel great.

I'll try this and and tweak things until they feel reasonable.

File: lib/cc/engine/violation_decorator.rb (hunk 3 above)
Comment (BlakeWilliams) [inline lib/cc/engine/violation_decorator.rb]: It just returns an array. It's super misleading and the alternative is to use named capture groups and `=~` which defines the capture group names as locals.

Not sure I have an opinion on which is better.

Comment (GordonDiggs) [inline lib/cc/engine/violation_decorator.rb]: Sorry - I wasn't clear in my comment. I mean is there a way to get the score and threshold from the violation object?

Comment (BlakeWilliams) [inline lib/cc/engine/violation_decorator.rb]: Ah, gotcha.

Unfortunately it seems like it only comes through `message`. https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/offense.rb

**Comment by BlakeWilliams** [comment]
@pbrisbin Updated to act more like the code you linked. The results I got from running it feel largely the same as before, which I assume is a good thing.

**Comment by pbrisbin** [comment]
@BlakeWilliams did you forget to push again?

**Comment by BlakeWilliams** [comment]
Yep, alias getting in the way. Removing it because it's clearly a recurring problem.

File: config/cops.yml [hunk 4]
```diff
@@ -1 +1,24 @@
-Metrics/ClassLength: 400_000
+Metrics/AbcSize:
+  violation_cost: 100_000
```
Comment (pbrisbin) [inline config/cops.yml]: This is my fault for not mentioning, but these are `_points` not `_cost` these days :)

Comment (BlakeWilliams) [inline config/cops.yml]: I was somewhat using them interchangeably. Do we prefer to rename these to points?

**Comment by pbrisbin** [comment]
Other than the old "cost" naming, this LGTM.

You can update the `.rubocopy.yml` to enforce the trailing-dot style instead. It's what we prefer anyway.

**Comment by BlakeWilliams** [comment]
I prefer it as well. Thanks for the review!

File: config/cops.yml (hunk 4 above)
Comment (pbrisbin) [inline config/cops.yml]: Definitely.

Both are terms and they mean different things. They're related as something like `cost = points / 100_000`.

Clearly, maintaining two values when one can be inferred from the other is kind of dumb, and we're on a pledge to only use "points" from now on.

**Pull request merged**